Core/Src/main.c

```diff
@@ -0,0 +1,595 @@
+
+//PID Control terms, adjust as needed
+int PropTerm = 10; //proportional term (30?)
+float IntTerm = 0; //Integral term
+float DerivTerm = 0.1; //Derivative term
+
+int FeedbackControlSignal; //final summation of effects of P, I and D controllers
+
+//Proportional control support variables
+int Pterm;
+
+//Integral control support variables
+int IntArray[30]; //= {0, 0, 0, 0, 0, 0, 0, 0, 0, 0};
+int IntegralTerm;
+int ITerm;
+int sum;
+
+//Derivative control support variables
+int DerArray[30]; // = {1, 1, 1, 1, 1, 1, 1, 1, 1, 1};
+int DerivativeTerm;
+int DTerm;
+int numerator;
+int denominator;
+int slope;
+
+
+int DesiredPos; //desired final position of the actuator
+
+int GoalTerm = 1; //sometimes overshoots, leading to undefined behavior. This prevents movement once goal is reached.
+int i;
+int j;
+int k;
+int f;
+int z;
+
+/* USER CODE END 0 */
+
+/**
+  * @brief  The application entry point.
+  * @retval int
+  */
+int main(void)
+{
+  /* USER CODE BEGIN 1 */
+
+  /* USER CODE END 1 */
+
+  /* MCU Configuration--------------------------------------------------------*/
+
+  /* Reset of all peripherals, Initializes the Flash interface and the Systick. */
+  HAL_Init();
+
+  /* USER CODE BEGIN Init */
+  Direction = 1; //Problems with global variables?
+  /* USER CODE END Init */
+
+  /* Configure the system clock */
+  SystemClock_Config();
+
+  /* USER CODE BEGIN SysInit */
+
+  /* USER CODE END SysInit */
+
+  /* Initialize all configured peripherals */
+  MX_GPIO_Init();
+  MX_USART2_UART_Init();
+  MX_TIM2_Init();
+  /* USER CODE BEGIN 2 */
+  HAL_TIM_Base_MspInit(&htim2);
+  HAL_TIM_Base_Start(&htim2); //Starts the TIM Base generation
+
+  HAL_TIM_PWM_Start(&htim2, TIM_CHANNEL_1);
+
+  __HAL_TIM_SET_COMPARE(&htim2, TIM_CHANNEL_1, DUTY_CYCLE);
+
+  HAL_TIM_MspPostInit(&htim2);
+  /* USER CODE END 2 */
+
+  /* Infinite loop */
+  /* USER CODE BEGIN WHILE */
+  while (1)
+  {
+    /* USER CODE END WHILE */
+
+    /* USER CODE BEGIN 3 */
+
+	  //PA15 motor pwm, PC2 And PC3 motor binary direction.
+	  //PA9 and PA10 are incremental motor feedback
+
+	  //MotorDirectionDetermination();
+
+	  DesiredPos = 8669;  //give offset (360 deg ~ 8669)
+
+	  errorterm = duration - DesiredPos;
+
+	  //control (P control)
+	  Pterm = PropTerm*errorterm;
+
+	  //Integral control (I control)
+	  ITerm = IntegralControl(errorterm);
+
+	  //Derivative control (D control)
+	  DTerm = DerivativeControl(errorterm);
+
+	  //Summing effects of PID controller
+	  FeedbackControlSignal = Pterm+ITerm+DTerm;
+
+	  if (FeedbackControlSignal > 60000 || FeedbackControlSignal < -60000){
+		  FeedbackControlSignal = 60000;
+	  }
+
+
+	  if(errorterm > 120) { //24 counts per degree
+		 HAL_GPIO_WritePin(GPIOC, GPIO_PIN_2, 1); //Forward or CW (Negative feedback)
+		 HAL_GPIO_WritePin(GPIOC, GPIO_PIN_3, 0);
+		 __HAL_TIM_SET_COMPARE(&htim2, TIM_CHANNEL_1, abs(FeedbackControlSignal));
+		 HAL_Delay(10);
+	  }
+	  if(errorterm < -120) {
```

Comment:
Use macros to define these. Looking at this I don't know where this 120 number comes from.